Review

konopat opened the pull request.

konopat committed:
test: добавлены e2e тесты с Cypress

- Добавлены тесты для модального окна ингредиента (открытие/закрытие)
- Добавлены тесты для добавления ингредиентов в конструктор
- Добавлены тесты для создания заказа
- Настроено тестовое окружение Cypress
- Добавлены моковые данные для тестов

konopat committed:
build: установка и настройка Cypress

- Установлен Cypress как dev-зависимость
- Добавлена базовая конфигурация cypress.config.js
- Настроены вспомогательные файлы в cypress/support
- Добавлены скрипты запуска в package.json

konopat committed:
test: добавлены unit тесты для Redux store

- Добавлены тесты для проверки инициализации rootReducer
- Добавлены тесты для constructor-slice (добавление, удаление, перемещение)
- Добавлены тесты для ingredients-slice (загрузка данных)
- Добавлены тесты для user-slice (авторизация)
- Добавлены тесты для orders-slice (создание заказа)
- Настроена конфигурация Jest
- Добавлены необходимые моки

SergeyK2004 commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,161 @@
+describe('Конструктор бургера', () => {
+  beforeEach(() => {
+    // Перехватываем запрос к API и возвращаем моковые данные
+    cy.intercept('GET', '**/api/ingredients', { fixture: 'ingredients.json' })
```
Отлично:
Хорошо, что используете intercept

SergeyK2004 commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,161 @@
+    it('должно открываться при клике на ингредиент', () => {
+      // Проверяем содержимое модального окна
+      cy.contains('Детали ингредиента').should('exist')
+      cy.contains('Краторная булка N-200i').should('exist')
```
Обратите внимание, что такой текст был найден на странице и без модального окна в 16 строке, т.е. проверка такого рода не подтверждает, что этот текст находится в модальном окне. Будет корректнее найти элемент модального окна и уже в нем выполнять поиск текста

SergeyK2004 commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,161 @@
+      .click()
+
+    // Проверяем, что булка добавилась в конструктор
+    cy.get('[class*="constructor-element"]')
```
Проверка в 56 строке подтверждает что такой элемент присутствовал на странице и раньше, возможно он как раз был в конструкторе, а не добавился позже.

Перед тем, как добавлять ингредиент, нужно проверить, что его нет в конструкторе. В противном случае вы не проверяете действие, а проверяете только наличие ингредиента в конструкторе, но добавился он из-за клика в кнопку "добавить" или уже был там неизвестно.
При проверке любого действия необходимо выполнить три шага:

1. Проверить, что элемент, который должен появиться в результате действия отсутствует.
2. Выполнить действие.
3. Проверить, что элемент появился в разметке.

Необходимо учесть это во всех проверках.

SergeyK2004 commented on cypress/e2e/constructor.cy.tsx:
```diff
@@ -0,0 +1,161 @@
+
+    // Подставляем токены авторизации
+    window.localStorage.setItem('accessToken', 'test_token')
+    window.localStorage.setItem('refreshToken', 'test_refresh_token')
```
Будет здорово, если после выполнения тестов моковые данные из хранилища и куков будут очищены

SergeyK2004 commented on src/services/store/slices/__tests__/constructor-slice.test.ts:
```diff
@@ -0,0 +1,68 @@
+import constructorReducer, {
```
Отлично:
Хорошо, что все тесты вынесены в отдельную папку.

SergeyK2004 commented on src/services/store/__tests__/store.test.ts:
```diff
@@ -0,0 +1,45 @@
+import store from '../store'
+
+describe('store', () => {
+  it('должен иметь корректное начальное состояние', () => {
```
**Надо исправить:**

Проверку корневого редьюсора нужно выполнить в соответствии с вашим чек-листом https://disk.yandex.ru/d/VRoqpskRKRVXqg

Вам нужно проверить не начальное состояние, а состояние стора при выполнении указанного в задании экшена.

SergeyK2004 reviewed (changes requested):
Очень хорошая работа. 
Тесты написаны аккуратно без лишнего кода. Код приятно читать. Все тесты проходят. 


Но есть небоольшие замечания, посмотрите их в коде, они позволят сделать ваши тесты надежнее. Так-же обратите внимание на тестирование корневого редьюсера. Этот тест нужно переделать в соответствии с заданием.

Commit:
test: улучшены e2e тесты конструктора бургера

- Добавлена корректная проверка состояния конструктора перед добавлением ингредиентов
- Реализованы все три шага проверки действия (проверка отсутствия -> действие -> проверка наличия)
- Исправлена проверка пустого конструктора в тесте создания заказа
- Обновлен тест корневого редьюсера для проверки неизвестного экшена

SergeyK2004 reviewed (approved):
Отличная работа.
Поздравляю со сдачей проекта.
Желаю успеха в дальнейшем обучении
